bundles/io/org.eclipse.smarthome.io.encryption/src/main/java/org/eclipse/smarthome/io/crypto/internal/CryptoModule.java

```diff
@@ -0,0 +1,137 @@
+
+    private int iterationCount = 2000;
+    private Cipher dcipher;
+    private Cipher ecipher;
+    private SecretKey secretKey;
+
+    public CryptoModule() {
+        initEnc();
+    }
+
+    private void initEnc() {
+        try {
+            byte[] encodedKey = CryptoActivator.getKey();
+            if (encodedKey != null) {
+                setEncPassphrase(encodedKey);
+                CryptoActivator.clearKey();
+            }
+        } catch (IllegalArgumentException e) {
+            logger.warn("Secret key initialization failed. ", e);
+        }
+    }
+
+    @Override
+    public String encrypt(String plainText) throws InvalidKeyException, IllegalBlockSizeException, BadPaddingException {
+        if (ecipher != null) {
+            try {
+                AlgorithmParameterSpec paramSpec = new PBEParameterSpec(salt, iterationCount);
+                ecipher.init(Cipher.ENCRYPT_MODE, secretKey, paramSpec);
+                byte[] plainTextBytes = plainText.getBytes("UTF-8");
+                byte[] encryptedBytes = ecipher.doFinal(plainTextBytes);
+                Base64.Encoder encoder = Base64.getEncoder();
+                String encryptedText = encoder.encodeToString(encryptedBytes);
+                return encryptedText;
+            } catch (UnsupportedEncodingException e) {
+                e.printStackTrace();
+            } catch (InvalidAlgorithmParameterException e) {
+                e.printStackTrace();
+            }
+            return null;
+
+        } else {
+            throw new IllegalStateException("Secret key not initialized");
+        }
+    }
+
+    @Override
+    public String decrypt(String encryptedText)
+            throws InvalidKeyException, IllegalBlockSizeException, BadPaddingException {
+
+        if (dcipher != null) {
+            try {
+                AlgorithmParameterSpec paramSpec = new PBEParameterSpec(salt, iterationCount);
+                dcipher.init(Cipher.DECRYPT_MODE, secretKey, paramSpec);
+                Base64.Decoder decoder = Base64.getDecoder();
+                byte[] encryptedTextBytes = decoder.decode(encryptedText);
+                byte[] decryptedBytes = dcipher.doFinal(encryptedTextBytes);
+                String decryptedText = new String(decryptedBytes, "UTF8");
+                return decryptedText;
+            } catch (UnsupportedEncodingException e) {
+                e.printStackTrace();
+            } catch (InvalidAlgorithmParameterException e) {
+                e.printStackTrace();
+            }
+            return null;
+        } else {
+            throw new IllegalStateException("Secret key not initialized");
+        }
+    }
+
+    @Override
+    public void setEncPassphrase(byte[] secret) {
+        try {
+            String text = new String(secret, "UTF-8");
+            KeySpec keySpec = new PBEKeySpec(text.toCharArray(), salt, iterationCount);
+            secretKey = SecretKeyFactory.getInstance("PBEWithMD5AndDES").generateSecret(keySpec);
```

Comment:
We definately do not want to use MD5 for security related hashes, nor DES to encrypt something. How about using PBEWithHmacSHA256AndAES_256 instead?

If PBEWithHmacSHA256AndAES_256 is not available and also with regards to US import/export restrictions we might want to have this configurable in a config file anyway so people with the JCE installed can use better algorithms.

Reply:
The java documentation explicitly mentions necessary algorithms to be shipped with every jre, so you should select the safest one as a fallback, if the configured one cannot be found. You might as well first try a stronger one, and catch the exception, before trying the jre guaranteed one.